File: kotlin-typescript/karakum/nameResolvers/resolveFunctionReturnTypePredicateName.js
import ts from "typescript";
import * as karakum from "karakum";

export default (node) => {
    if (
        node.parent
        && ts.isTypePredicateNode(node.parent)

        && node.parent.parent
        && ts.isFunctionDeclaration(node.parent.parent)
        && ts.isIdentifier(node.parent.parent.name)
        && (node.parent.parent.type === node.parent)
    ) {
        const parentName = node.parent.parent.name.text

        return `${karakum.capitalize(parentName)}ResultPredicate`
    }

    return null
}

File: kotlin-typescript/karakum/src/jsMain/kotlin/typescript/karakum/nameResolvers/resolveFunctionReturnTypePredicateName.kt
package typescript.karakum.nameResolvers

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.Context
import io.github.sgrishchenko.karakum.extension.plugins.TypeScriptService
import io.github.sgrishchenko.karakum.extension.plugins.typeScriptServiceKey
import typescript.Node
import typescript.isFunctionDeclaration
import typescript.isTypePredicateNode

fun resolveFunctionReturnTypePredicateName(node: Node, context: Context) = nullable {
    val typeScriptService = ensureNotNull(context.lookupService(typeScriptServiceKey))

    val typePredicate = ensureNotNull(typeScriptService.getParent(node))
    ensure(isTypePredicateNode(typePredicate))

    val function = ensureNotNull(typeScriptService.getParent(typePredicate))
    ensure(isFunctionDeclaration(function))
    ensure(function.type === typePredicate)
    val parentName = ensureNotNull(function.name).text

    "${parentName.replaceFirstChar { it.titlecase() }}ResultPredicate"
}
